ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -67,25 +183,529 @@ void end_heap_allocation_recording(struct heap_recorder *heap_recorder, ddog_pro
+  // num_tracked_objects reached 0
+  heap_record->num_tracked_objects++;
+  // Then update object_records
+  object_record_update_data update_data = (object_record_update_data) {
+    .heap_recorder = heap_recorder,
+    .new_object_record = object_record_new(obj, heap_record, object_data),
+  };
+  st_update(heap_recorder->object_records, obj, update_object_record_entry, (st_data_t) &update_data);
+}
+
+// Struct holding data required for an update operation on heap_records
+typedef struct {
+  // [in] The stack this heap_record update operation is associated with
+  // NOTE: Transfer of ownership is assumed, do not re-use it after call to ::update_heap_record_entry_with_new_allocation
+  heap_stack *stack;
+  // [out] Pointer that will be updated to the updated heap record to prevent having to do
+  // another lookup to access the updated heap record.
+  heap_record **record;
+} heap_record_update_data;
+
+// This function assumes ownership of stack_data is passed on to it so it'll either transfer ownership or clean-up.
+static int update_heap_record_entry_with_new_allocation(st_data_t *key, st_data_t *value, st_data_t data, int existing) {
+  heap_record_update_data *update_data = (heap_record_update_data*) data;
+
+  if (existing) {
+    // there's already a heap_record matching the stack, lets just re-use it which means
+    // we should free the stack
+    heap_stack_free(update_data->stack);
+  } else {
+    // no matching heap record, lets allocate new heap_record_key and heap_record based on the stack
+    // and transfer ownership to the hash
+    (*key) = (st_data_t) heap_record_key_new(update_data->stack);
+    (*value) = (st_data_t) heap_record_new(update_data->stack);
+  }
+
+  heap_record *record = (heap_record*) (*value);
+  (*update_data->record) = record;
+
+  return ST_CONTINUE;
+}
+
+// This version assumes that a new stack got created and there's an expectation of transfer of ownership of
+// it to this function. For stack re-use you should be using ::commit_allocation_with_heap_record.
+// WARN: Expects records_mutex to be held
+static void commit_allocation_with_heap_stack(heap_recorder *heap_recorder, heap_stack *heap_stack, VALUE obj, live_object_data object_data) {
+  // First lets update the heap_records with this stack info
+  // NOTE: Using a stack-allocated key here for easy cleanup logic. The update function will create a heap-allocated key
+  // if no matching heap_record exists.
+  heap_record_key lookup_key = (heap_record_key) {
+    .type = HEAP_STACK,
+    .heap_stack = heap_stack,
+  };
+  heap_record *heap_record = NULL;
+  heap_record_update_data update_data = (heap_record_update_data) {
+    .stack = heap_stack,
+    .record = &heap_record,
+  };
+  st_update(heap_recorder->heap_records, (st_data_t) &lookup_key, update_heap_record_entry_with_new_allocation, (st_data_t) &update_data);
+
+  commit_allocation_with_heap_record(heap_recorder, heap_record, obj, object_data);
```

Comment:
Hmm yeah I see what you mean. It was split like this because queue logic was another place from which we could call `commit_allocation_with_heap_stack`. Given what I mentioned above in the other comment, we may be able to simplify indeed!